Add Interchange project update (#373)
openforcefield/openforcefield.org: content/community/news/general/2021-11-10-interchange-update.md

    Torsion           0.000069
    vdW               0.016327
    Electrostatics    0.046856
    dtype: float64

Are these errors on the order of 10<sup>-4</sup> - 10<sup>-2</sup> kJ/mol between engines'
evaluation of supposedly-equivalent inputs acceptable? That's an [open
question](https://dx.doi.org/10.1007%2Fs10822-016-9977-1) - and in this case, it's likely that the
``Interchange``'s exports to each have some small discrepancies and how each engine is called by its
driver is not entirely precise. Regardless, this is a useful framework for ensuring outputs to these
engines, and engines that may be supported in the future, are consistent with each other. This also
serves as a template for how reference energies could be distributed by force field developers:
given a molecule topology prepared in a perscribed way, force field XYZ should provide these
particular energies. At present, we have a decent idea of what engines _think_ these energies are
but not what the original developers intended them to be; reference energies would enable a more
direct route to testing engines and perhaps uncover bugs (relatively minor, but perhaps
non-negligible) in engines themselves.

> ## Python package

Author: I'm not 100% sure this section is worth the text it takes up; it serves two purposes, but I'm not sure it serves them well.
1. Try to wrap up some of the topics covered earlier and connect the functionality back to the source code
2. Provide a high-level description of the intent of each module; I have some feedback from adjacent developers that tried to poke around in the source code but got lost quickly, "you have the components, and then drivers, and then interop, and I'm not sure where stuff actually lives." Arguably this should be addressed in the documentation itself

Reviewer: I think this is great to keep in the post. The contents of a blog post are "whatever you want to write about". If this section was the entire post, that would also be fine. As you say in point 2, docs are where users should go for a concise, structured explanation of specific topics, and we're free to use material from this blog post as a starting point for a page in the docs.

---
date: "2021-11-10T00:00:00+00:00"
title: 'Interchange project update'
tags: ["infrastructure", "interoperability"]
categories: ["general"]
draft: false
description: 'Interchange: Enabling the broader use of OpenFF force fields'
slug: 'interchange-update-2021-11-10'
weight: 10
markup: markdown
thumb: "foo.png"
author: "Matt Thompson"
---

Thus far, the only simulation engine natively supported by [export from the OpenFF software
stack](https://open-forcefield-toolkit.readthedocs.io/en/latest/api/generated/openff.toolkit.typing.engines.smirnoff.forcefield.ForceField.html#openff.toolkit.typing.engines.smirnoff.forcefield.ForceField.create_openmm_system)

> is been OpenMM. OpenMM offers great performance on GPUs and excellent flexibility via its API, but
> many scientists wish to use tools that are only compatible with other engines. There are tools that
> can convert OpenMM objects to file formats understood by other engines, but these steps can have
> rough edges and add complexity to users' workflows. The object of the Interchange project is to
> provide a first-class support for these operations with the OpenFF software stack. This enabled by
> a Python package of the same name that is described below and in its
> [documentation](https://openff-interchange.readthedocs.io/en/latest/index.html#).

Reviewer: ```suggestion
has been OpenMM. OpenMM offers great performance on GPUs and excellent flexibility via its API, but
many scientists wish to use tools that are only compatible with other molecular simulation engines, like AMBER and GROMACS. There are tools that
can convert OpenMM objects to file formats understood by these other engines, but these steps can have
rough edges and add complexity to users' workflows. For this reason, OpenFF has started the Interchange project. The objective of the Interchange project is to
provide a first-class support for these operations with the OpenFF software stack. This enabled by
a Python package of the same name that is described below and in its
[documentation](https://openff-interchange.readthedocs.io/en/latest/index.html#).
```

> This post aims to provide a high-level overview of the aims of the project, its current status,

Reviewer: ```suggestion
This post aims to provide a high-level overview of the aims of the Interchange project, its current status,
```

present capabilities, and some future directions.

{{< note >}}
Like much of of OpenFF infrastructure, Interchange is rapidly evolving and details within this blog post are likely to be out of date in the future.
{{< /note >}}

Code in this post should run if copy-pasted into a Jupyter notebook or Python interpreter. This is
being written in November 2021 and code snippets intended to be used with Interchange v0.1.3. The API is
not yet stable; the names, import paths, and structure of key components will likely be updated in
the future in accordance with user feedback. If you'd like to follow along, see installation
instructions at the bottom of the page; ``Interchange`` is distributed under the MIT license and all
functionality is accessible using only free and open-source tools.

## Contents of an Interchange object

To emphasize a detail above, the key objective of the Interchange project at present is to make it
easier to get OpenFF force fields in use in non-OpenMM workflows. The scope of this task is large,
but it is finite; the

> [SMIRNOFF](https://open-forcefield-toolkit.readthedocs.io/en/0.10.1/users/smirnoff.html)

Reviewer: Should point to standards repo since that's the new authoritative home of the SMIRNOFF spec

specification used by OpenFF force fields provides clarity to many questions of information
content in force fields, which is key for developing an interface to it. The [OpenFF
Toolkit](https://open-forcefield-toolkit.readthedocs.io/en/0.10.1/topology.html) also already provides
convenient representations of molecular topologies and functions for parsing existing data into its
models. Note that this does not close the door on interfacing with non-OpenFF models, it simply
provides a starting point. Expanding this work outside of the OpenFF stack is discussed below
("Future work").

An `Interchange` object stores all the information known about a system; this includes its
chemistry, how that chemistry is represented by a force field, and how the system is organised in 3D
space. An `Interchange` object has four components, each of which is also an attribute of the core
Python class

The `topology` attribute stores chemical information in a graph-like structure, without specific information
about forces from the force field. Currently, this attribute is literally a

> [`Topology`](https://open-forcefield-toolkit.readthedocs.io/en/latest/api/generated/openff.toolkit.topology.Topology.html#openff.toolkit.topology.Topology)

Reviewer: Other RTD links go to `0.10.1`, but this one goes to `latest` - I don't have a strong opinion on which to point to, but it'd be good to standardize throughout the blog post.

Author: I've chosen to explicitly link to the current versions. I can easily be convinced otherwise if Josh disagrees while reviewing

> object from the OpenFF Toolkit. This representation includes necessary information for molecular

Reviewer: ```suggestion
object from the OpenFF Toolkit. The OpenFF Toolkit Topology object includes necessary information for molecular
```

mechanics representations (masses, elements, and basic bond information) but also richer chemical
information used by cheminformatics toolkits and some portions of the SMIRNOFF specification (bond
orders, bond and atom stereochemistry, aromaticity). While this may change in the future, this does
mean that coarse-grained representations are not supported in any explicit way.

The `handlers` attribute is a mapping between chemical information and force field parameters as
they are applied to the system (i.e. post-parametrization). A unique identifier of a segment of the
chemical graph - for example, the four atoms participating in a dihedral - is mapped to a unique
identifier of the physical parameters assigned to them - in this example, the force constant, phase,
and periodicity of the dihedral force. This allows for some de-duplication of parameters; by
pointing to a representation of the force, each unique parameter is only stored once. The mapping
between chemical topology and force parameters serves as a way of connecting the forces actually
present in a simulation to how they were applied by the force field.

> Particle positions are stored as a unit-tagged `Nx3` matrix, where `N` is the number of particles in

Reviewer: I had trouble spotting that positions and box vectors were items 3 and 4 of the "four components" - Maybe some additional formatting could clear this up.

For more complex systems, there may be other steps involved at different points in a workflow, such
as adding solvent and counter-ions or combining multiple ``Interchange`` objects generated from
different force fields. But for this example, we're (partially) done! The ``Interchange`` object is
fully constructed and we can export its contents to objects understood by simulations engines.

As a brief aside,
[migrating](https://openff-interchange.readthedocs.io/en/latest/using/migrating.html#migrating-from-the-openff-toolkit)
existing code from using the OpenFF Toolkit's OpenMM interface to using Interchange directly is
straightforward. Wherever
[`ForceField.create_openmm_system](https://open-forcefield-toolkit.readthedocs.io/en/latest/api/generated/openff.toolkit.typing.engines.smirnoff.forcefield.ForceField.html#openff.toolkit.typing.engines.smirnoff.forcefield.ForceField.create_openmm_system)
is was used, simply replace it with `Interchange.from_smirnoff` like so:

```python
# Existing code generating an OpenMM System
openmm_system = forcefield.create_openmm_system(topology)

# New code generating an Interchange object

> out = Interchange.from_smirnoff(forcefield, topology)

Reviewer: Maybe have the variable set by this line be called `interchange` to make it less ambiguous

``openff.interchange.drivers`` module that each take in an ``Interchange`` object, call out to an
engine a single-point energy calculation, and store the result.

This snippet continues from above, since we have conveniently already done the work of constructing
our object. Note that these functions are relatively unintelligent at the moment and simply look to
see if each engine is installed (i.e. if `gmx`, `sander`, etc. are in `$PATH`). These might not run
on a machine that does not have them installed - feel free to skip lines if you're following along.

```python
from openff.interchange.drivers.amber import get_amber_energies
from openff.interchange.drivers.openmm import get_openmm_energies
from openff.interchange.drivers.gromacs import get_gromacs_energies
from openff.interchange.drivers.lammps import get_lammps_energies
```

```python
print(get_gromacs_energies(out))
```
    Energies:

    Bond:          		17.31004524230957 kJ / mol
    Angle:         		195.2955780029297 kJ / mol
    Torsion:       		13.520307540893555 kJ / mol
    Nonbonded:     		None
    vdW:           		43.82065216824412 kJ / mol
    Electrostatics:		-707.1108989715576 kJ / mol

> 

Reviewer: ```suggestion

(Note that all of these outputs will have a `None` for their nonbonded energy, since the underlying engines have separated this out into `vdW` and `Electrostatics` components already)

```

    GROMACS  17.310045  195.295578  13.520308  43.820652     -707.110899
    LAMMPS   17.309886  195.295355  13.520323  43.783375     -707.013998

This allows us to quickly look see how much each engine tends to differ in its computation of each
component of the potential energy function:

```python3
summary_data.std()  # or .describe(), etc.
```
    Bond              0.000100
    Angle             0.000121
    Torsion           0.000069
    vdW               0.016327
    Electrostatics    0.046856
    dtype: float64

Are these errors on the order of 10<sup>-4</sup> - 10<sup>-2</sup> kJ/mol between engines'
evaluation of supposedly-equivalent inputs acceptable? That's an [open
question](https://dx.doi.org/10.1007%2Fs10822-016-9977-1) - and in this case, it's likely that the
``Interchange``'s exports to each have some small discrepancies and how each engine is called by its
driver is not entirely precise. Regardless, this is a useful framework for ensuring outputs to these
engines, and engines that may be supported in the future, are consistent with each other. This also
serves as a template for how reference energies could be distributed by force field developers:

> given a molecule topology prepared in a perscribed way, force field XYZ should provide these

Reviewer: I've reached the limits of my knowledge of the English language. Should the be `prescribed` or `proscribed`? 

```suggestion
given a molecule topology prepared in a prescribed way, force field XYZ should provide these
```

particular energies. At present, we have a decent idea of what engines _think_ these energies are
but not what the original developers intended them to be; reference energies would enable a more
direct route to testing engines and perhaps uncover bugs (relatively minor, but perhaps
non-negligible) in engines themselves.

## Python package

> This project is hosted in a Python package (`openff-interchange`) which store the actual `Interchange` class. Development occurs in public on a [GitHub repository](https://github.com/openforcefield/openff-interchange). The package itself is made up of pieces that have looked in the above sections. To a rough approximation, let's just look at some of the files and folders in the module:

Reviewer: ```suggestion
This project is hosted in a Python package (`openff-interchange`) which stores the actual `Interchange` class. Development occurs in public on a [GitHub repository](https://github.com/openforcefield/openff-interchange). The package itself is made up of pieces that have looked in the above sections. To a rough approximation, let's just look at some of the files and folders in the module:
```

```shell
$ tree -L 1 openff/interchange
openff/interchange
├── __init__.py
├── _version.py
├── components  # Core internal modules
├── conftest.py
├── data
├── drivers
├── energy_tests
├── exceptions.py
├── interop
├── interoperability_tests
├── models.py
├── py.typed
├── testing
├── types.py
├── unit_tests
└── utils.py
```

Many of these files and folers are for testing and/or Python-specific configuration, but some are the source of what we've looked at today. `components/` stores the key models used internally and the `Interchange` class itself, `/drivers/` stores the functions used in energy-based tests, and `interop/` stores the code used to actually _do_ each export.

## Future work

Development is currently underway on features that will enable the typing and parametrization of
biopolymers with SMIRNOFF-style force fields. This work is progressing alongside a significant

> refactor to the OpenFF toolkit to support biopolymers in its `Molecule` and `Topology` models. This
> work is undergoing extensive internal testing as of November 2021 and we hope to have it in a

Reviewer: ```suggestion
refactor to the OpenFF toolkit to support biopolymers in its `Molecule` and `Topology` models. The 
refactor is undergoing extensive internal testing as of November 2021 and we hope to have it in a
```

pre-release build of the OpenFF Toolkit in early 2021. This will enable Interchange to work with the
rest of the OpenFF stack on biopolymers, supporting our scientists currently working toward the first protein
force field produced by the initiative and external scientists who will use it in the future.

Surveys of OpenFF's user base indicate that most people use OpenMM, Amber, and/or GROMACS as their
engine of choice, so compatibility with those engines has been the focus of development. (Support
for LAMMPS was added on the basis of my experience with it.) In general, the current focus is on
improving the reliability and feature parity of the engines that are currently support.  On this
basis, support for other _exports_ is not highly-prioritized at the moment. If support for other
engines would enable the use of OpenFF force fields in your workflows, we'd be happy to hear from
you! (See "Feedback" at the end of this post).

> The [SMIRNOFF](https://open-forcefield-toolkit.readthedocs.io/en/0.10.1/users/smirnoff.html)

Reviewer: Ditto above, should point to `standards` repo

is not currently prioritized due to relatively little scientific interest at the moment.

Lastly, support for _importing_ from other object models is planned in general terms but is not a
topic of active development at this moment. As a proof of concept that this is possible - that
``Interchange`` is not wholly specific to SMIRNOFF force fields and the OpenFF stack - we worked
with the [MoSDeF](https://mosdef.org) team at Vanderbilt University to add
[``Interchange.from_foyer``](https://openff-interchange.readthedocs.io/en/latest/_autosummary/openff.interchange.components.interchange.Interchange.html#openff.interchange.components.interchange.Interchange.from_foyer), which, as the name implies, creates an ``Interchange`` object from a
[Foyer](foyer.mosdef.org) force field and a minimal chemical graph. Note that the structure of
the function is similar (force field plus chemical topology) but with different object swapped in!
We hope that this can work be extended to support some amount of _importing_ from existing force
field infrastructure, such as using ligands parametrized with OpenFF force fields with proteins

> parametrized with Amber force fields. We hope that ``Interchange`` has been engineering in a way
> that swapping out components is feasible not only for us but also other developers.

Reviewer: ```suggestion
parametrized with Amber force fields. We hope that ``Interchange`` has been engineered in a way
that makes swapping out components feasible not only for us but also other developers.
```
